source/module_hsolver/diago_elpa_native.cpp

```diff
@@ -0,0 +1,150 @@
+#include "module_base/scalapack_connector.h"
+#include "module_base/timer.h"
+#include "module_base/tool_quit.h"
+#include "module_hsolver/genelpa/elpa_new.h"
+#include "omp.h"
+
+namespace hsolver
+{
+#ifdef __MPI
+template <typename T>
+MPI_Comm DiagoElpaNative<T>::setmpicomm()
+{
+    if (this->elpa_num_thread == -1)
+    {
+        return MPI_COMM_WORLD;
+    }
+    else
+    {
+        int _num = 0;
+        MPI_Comm_size(MPI_COMM_WORLD, &_num);
+        if (elpa_num_thread > _num || elpa_num_thread <= 0)
+        {
+            return MPI_COMM_WORLD;
+        }
+        else
+        {
+            lastmpinum++;
+            int* _ranks = new int[elpa_num_thread];
+            for (int i = 0; i < elpa_num_thread; i++)
+            {
+                _ranks[i] = (lastmpinum + i) % _num;
+            }
+            MPI_Group _tempgroup, _oldgroup;
+            MPI_Comm_group(MPI_COMM_WORLD, &_oldgroup);
+            MPI_Group_incl(_oldgroup, elpa_num_thread, _ranks, &_tempgroup);
+            MPI_Comm _new_comm;
+            MPI_Comm_create(MPI_COMM_WORLD, _tempgroup, &_new_comm);
+            delete[] _ranks;
+            return _new_comm;
+        }
+    }
+}
+#endif
+
+#ifdef __MPI
+template <typename T>
+void DiagoElpaNative<T>::diag_pool(hamilt::MatrixBlock<T>& h_mat,
+                                   hamilt::MatrixBlock<T>& s_mat,
+                                   psi::Psi<T>& psi,
+                                   Real* eigenvalue_in,
+                                   MPI_Comm& comm)
+{
+    std::vector<Real> eigen(GlobalV::NLOCAL, 0.0);
+
+    ModuleBase::timer::tick("DiagoElpaNative", "elpa_solve");
+
+    int nev = GlobalV::NBANDS;
+    int narows = h_mat.row;
+    int nacols = h_mat.col;
+
+    int cblacs_ctxt = h_mat.desc[1];
+    int nFull = h_mat.desc[2];
+    int nblk = h_mat.desc[4];
+    // cout<<"parameters are passed\n";
+    int nprows, npcols, myprow, mypcol;
+
+    Cblacs_gridinfo(cblacs_ctxt, &nprows, &npcols, &myprow, &mypcol);
+
+    if (elpa_init(20210430) != ELPA_OK)
+    {
+        fprintf(stderr, "Error: ELPA API version not supported");
+    }
+
+    // elpa_init(20210430);
+    int success;
+    elpa_t handle = elpa_allocate(&success);
+    int num_threads = omp_get_max_threads();
+    elpa_set(handle, "omp_threads", num_threads, &success);
+    elpa_set(handle, "na", (int)nFull, &success);
+    elpa_set(handle, "nev", (int)nev, &success);
+    elpa_set(handle, "local_nrows", (int)narows, &success);
+    elpa_set(handle, "local_ncols", (int)nacols, &success);
+    elpa_set(handle, "nblk", (int)nblk, &success);
+    elpa_set(handle, "mpi_comm_parent", (int)(MPI_Comm_c2f(comm)), &success);
+    elpa_set(handle, "process_row", (int)myprow, &success);
+    elpa_set(handle, "process_col", (int)mypcol, &success);
+    elpa_set(handle, "blacs_context", (int)cblacs_ctxt, &success);
+    elpa_setup(handle);
+    elpa_set(handle, "solver", ELPA_SOLVER_1STAGE, &success);
+
+#ifdef __CUDA
```

Comment:
Maybe it is possible to use `ELPA_WITH_NVIDIA_GPU_VERSION` from [elpa/elpa_configured_options.h](https://gitlab.mpcdf.mpg.de/elpa/elpa/-/blob/master/elpa/elpa_configured_options.h.in) to determine if elpa is installed with GPU support.